server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java

```diff
@@ -0,0 +1,110 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.vault;
+
+import javax.mail.MessagingException;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailrepository.api.MailKey;
+import org.apache.james.mailrepository.api.MailRepository;
+import org.apache.james.mailrepository.api.MailRepositoryStore;
+import org.apache.james.mailrepository.api.MailRepositoryUrl;
+import org.apache.james.util.streams.Iterators;
+import org.reactivestreams.Publisher;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class MailRepositoryDeletedMessageVault implements DeletedMessageVault {
+    public static class Configuration {
+        private final MailRepositoryUrl urlPrefix;
+
+        public Configuration(MailRepositoryUrl urlPrefix) {
+            this.urlPrefix = urlPrefix;
+        }
+    }
+
+    private final MailRepositoryStore mailRepositoryStore;
+    private final Configuration configuration;
+    private final MailConverter mailConverter;
+
+    public MailRepositoryDeletedMessageVault(MailRepositoryStore mailRepositoryStore, Configuration configuration, MailConverter mailConverter) {
+        this.mailRepositoryStore = mailRepositoryStore;
+        this.configuration = configuration;
+        this.mailConverter = mailConverter;
+    }
+
+    @Override
+    public Publisher<Void> append(User user, DeletedMessage deletedMessage) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(deletedMessage);
+
+        MailRepository mailRepository = repositoryForUser(user);
+
+        return Mono.fromCallable(() -> mailRepository.store(mailConverter.toMail(deletedMessage)))
+            .publishOn(Schedulers.elastic())
+            .then();
+    }
+
+    @Override
+    public Publisher<Void> delete(User user, MessageId messageId) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(messageId);
+
+        MailRepository mailRepository = repositoryForUser(user);
+        MailKey mailKey = new MailKey(messageId.serialize());
+
+        return Mono.fromRunnable(Throwing.runnable(() -> mailRepository.remove(mailKey)))
+            .publishOn(Schedulers.elastic())
+            .then();
+    }
+
+    @Override
+    public Publisher<DeletedMessage> search(User user, Query query) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(query);
+        Preconditions.checkArgument(query.getCriteria().isEmpty(), "Search is not supported yet...");
```

Comment:
yet you can apply the query on the Flux, isn't it?

Reply:
No. The ability to filter flux based on a query is done in #2193 and is not contributed yet.

Reply:
ok